src/ray/raylet/node_manager.cc

```diff
@@ -519,36 +553,35 @@ void NodeManager::HandleActorStateTransition(const ActorID &actor_id,
   }
 
   if (actor_registration.GetState() == ActorState::ALIVE) {
-    // The actor's location is now known. Dequeue any methods that were
-    // submitted before the actor's location was known.
-    // (See design_docs/task_states.rst for the state transition diagram.)
-    const auto &methods = local_queues_.GetTasks(TaskState::WAITING_FOR_ACTOR_CREATION);
-    std::unordered_set<TaskID> created_actor_method_ids;
-    for (const auto &method : methods) {
-      if (method.GetTaskSpecification().ActorId() == actor_id) {
-        created_actor_method_ids.insert(method.GetTaskSpecification().TaskId());
-      }
-    }
-    // Resubmit the methods that were submitted before the actor's location was
-    // known.
-    auto created_actor_methods = local_queues_.RemoveTasks(created_actor_method_ids);
-    for (const auto &method : created_actor_methods) {
-      if (!lineage_cache_.RemoveWaitingTask(method.GetTaskSpecification().TaskId())) {
-        RAY_LOG(WARNING) << "Task " << method.GetTaskSpecification().TaskId()
-                         << " already removed from the lineage cache. This is most "
-                            "likely due to reconstruction.";
-      }
-      // Maintain the invariant that if a task is in the
-      // MethodsWaitingForActorCreation queue, then it is subscribed to its
-      // respective actor creation task. Since the actor location is now known,
-      // we can remove the task from the queue and forget its dependency on the
-      // actor creation task.
-      RAY_CHECK(task_dependency_manager_.UnsubscribeDependencies(
-          method.GetTaskSpecification().TaskId()));
-      // The task's uncommitted lineage was already added to the local lineage
-      // cache upon the initial submission, so it's okay to resubmit it with an
-      // empty lineage this time.
-      SubmitTask(method, Lineage());
+    // The actor's location is now known.
+    bool resumed_from_checkpoint = checkpoint_id_to_restore_.count(actor_id) > 0;
+    if (!resumed_from_checkpoint) {
+      // If the actor wasn't resumed from a checkpoint, just resubmit the waiting tasks.
+      ResubmitTasksWaitingForActorCreation(actor_id);
+    } else {
+      // If this actor was resumed from a checkpoint, look up the checkpoint in GCS,
+      // retore actor state, and resubmit the waiting tasks.
+      const auto checkpoint_id = checkpoint_id_to_restore_[actor_id];
```

Comment:
Hmm I'm not sure if this will work in a distributed setting, since the publish could still go out to other nodes, which will resubmit their cached tasks, and that could potentially happen before we restore the checkpoint here.

Reply:
Ah, yes, that problem could still happen.